### ISSUE TITLE
[Potential NaN bug] Loss may become NaN during training

### ISSUE THREAD

**Issue by Justobe** [winnietsang/handson-tensorflow#2]
Hello~

Thank you very much for sharing the code!

I try to use my own data set ( with the same shape as mnist) in  code. After some iterations, it is found that the training loss becomes NaN. After carefully checking the code, I found that the following code may trigger NaN in loss:

In `handon-tensorflow/code-10-image.py:  `

```python
cross_entropy = tf.reduce_mean(-tf.reduce_sum(y_ * tf.log(y), reduction_indices=[1]), name='cross_entropy')
```

If  y contains 0 (output of softmax ), the result of `tf.log(y)` is `inf` because `log(0)` is illegal . And this may cause the result of loss to become NaN.

It could be fixed by making the following changes:

```python
cross_entropy = tf.reduce_mean(-tf.reduce_sum(y_ * tf.log(y + 1e-10), reduction_indices=[1]), name='cross_entropy')
```

or

```python
cross_entropy = tf.reduce_mean(-tf.reduce_sum(y_ * tf.log(tf.clip_by_value(y,1e-10,1.0)), reduction_indices=[1]), name='cross_entropy')
```

Hope to hear from you ~

Thanks in advance!  : )
